### PR TITLE
add ltproperties

### PR DESCRIPTION
**Pull request by u-fischer**
# Internal housekeeping

## Status of pull request

- Ready to merge

## Checklist of required changes before merge will be approved
- [x] Test file(s) added
- [x] Version and date string updated in changed source files
- [x] Relevant `\changes` entries in source included
- [x] Relevant `changes.txt` updated
- [x] `ltnewsX.tex` (and/or `latexchanges.tex`) updated

### CODE REVIEW

File: base/doc/ltnews38.tex [hunk 1]
```diff
@@ -148,6 +148,62 @@ \section{Sockets and plugs}
...
+The new module predeclares a set of generally useful properties: 
+\texttt{currentlabel} (the standard counter representation), \texttt{page} 
+(the page representation), \texttt{title} (the title if set e.g. by 
+\pkg{nameref}), \texttt{target} (the name of the link target if set e.g. by 
+\pkg{hyperref}), \texttt{pagenum} (the page as an arabic number), 
+\texttt{abspage} (the absolute page number), \texttt{counter} (the name of 
+the current counter), \texttt{xpos} and \texttt{ypos} the position on the 
+page as set by a previous \cs{pdfsavepos}. 
+  
+The module provides both \LaTeX2e camel case commands and expl3 commands.
```
Comment (josephwright) [inline base/doc/ltnews38.tex:204]: `\LaTeXe{}`

File: base/ltproperties.dtx [hunk 2]
```diff
@@ -0,0 +1,930 @@
...
+%   \pkg{hyperref} and gives the name of the last destination it created. 
+% \end{variable}
+% 
+% \begin{variable}{counter}
+%   (now) The content of \cs{@currentcounter}. 
+%   This command contains after a \cs{refstepcounter} the name of the counter.
+% \end{variable}
+% 
+% \begin{variable}{xpos,ypos}
+%   (shipout)  This stores the x and y coordinates of a point previously
```
Comment (josephwright) [inline base/ltproperties.dtx:422]: `$x$~and $y$`

File: base/ltproperties.dtx [hunk 3]
```diff
@@ -0,0 +1,930 @@
...
+% 
+% \begin{variable}{counter}
+%   (now) The content of \cs{@currentcounter}. 
+%   This command contains after a \cs{refstepcounter} the name of the counter.
+% \end{variable}
+% 
+% \begin{variable}{xpos,ypos}
+%   (shipout)  This stores the x and y coordinates of a point previously 
+%   stored with \cs{pdfsavepos}/\cs{savepos}. 
+%   E.g. (if bidi is used it can be necessary to save the position
```
Comment (josephwright) [inline base/ltproperties.dtx:424]: Tie

File: base/ltproperties.dtx [hunk 4]
```diff
@@ -0,0 +1,930 @@
...
+%   the default and a flag to show if the code works now or at shipout. This
+%   last entry is done using text so needs a check. We could use a set of
+%   |prop| here, but as we never need to map or copy the lists, we can gain
+%   performance using the hash table approach.
+%    \begin{macrocode}
+\cs_new_protected:Npn \property_new:nnnn #1#2#3#4
+  {
+    \cs_if_free:cTF { @@_name_ #1 : }
+      {
+        \exp_args:Nx \@@_gset:nnnn { \tl_to_str:n {#1} }
```
Comment (josephwright) [inline base/ltproperties.dtx:479]: Probably should declare a variant here and use `\@@_gset:xnnn`

File: base/ltproperties.dtx [hunk 5]
```diff
@@ -0,0 +1,930 @@
...
+%
+% \begin{macro}{\IfLabelExistTF}
+%    \begin{macrocode}
+\cs_new_eq:NN \IfLabelExistTF \property_if_recorded:nTF  
+%    \end{macrocode}
+% \end{macro} 
+%
+% \begin{macro}[pTF]{\property_if_recorded:nn}
+% tests if the label/property combination has been set
+% This can then be used to setup e.g. rerun messages.
```
Comment (josephwright) [inline base/ltproperties.dtx:749]: Tie

File: base/ltproperties.dtx [hunk 6]
```diff
@@ -0,0 +1,930 @@
...
+%    \end{macrocode}
+% \end{macro}
+% 
+% \begin{macro}{\property_ref_undefined_warn:}
+%    \begin{macrocode}
+\cs_new_protected:Npn \property_ref_undefined_warn:
+ {
+   \flag_if_raised:nT { l_@@_ref_flag }
+    {
+      \G@refundefinedtrue
```
Comment (josephwright) [inline base/ltproperties.dtx:777]: Should this be a `\legacy_if_set_true:n`?

Comment (josephwright) [reply base/ltproperties.dtx:777]: (Also applies to the other uses)

File: base/ltproperties.dtx [hunk 7]
```diff
@@ -0,0 +1,930 @@
...
+%    \begin{macrocode}
+\property_new:nnnn  { xpos } { shipout } { 0} { \int_use:N \tex_lastxpos:D }
+\property_new:nnnn  { ypos } { shipout } { 0} { \int_use:N \tex_lastypos:D } 
+%    \end{macrocode}
+% \end{variable}
+%  
+%  \subsection{Messages}
+%    \begin{macrocode}
+\msg_new:nnnn { property } { exists }
+    { property~'#1'~ has~ already~ been~ declared. }
```
Comment (josephwright) [inline base/ltproperties.dtx:874]: 'Property'

File: base/ltproperties.dtx [hunk 8]
```diff
@@ -0,0 +1,930 @@
...
+      name.\\
+      Please~ use~ a~ different~ name~ for~ your~ property.}
+      
+\msg_new:nnnn { property } { not-declared }
+  { Property~'#1'~not~declared. }
+  {
+    LaTeX~has~been~asked~to~use~property~'#1',~but~this~
+    name~has~not~been~declared.
+  }
+\msg_new:nnnn { property } { unknown-setpoint }
```
Comment (josephwright) [inline base/ltproperties.dtx:885]: Why use ` `#3`, not just use `#2`?

**Review by josephwright — CHANGES_REQUESTED** [review]
Very minor tweaks here: mainly just typos. (If I have time later today I'll likely pick off myself.)

File: base/ltproperties.dtx [hunk 9]
```diff
@@ -0,0 +1,930 @@
...
+\begin{document}
+  \DocInput{ltproperties.dtx}
+\end{document}
+%</driver>
+% \fi
+%
+% \title{^^A
+%   Recording and cross-referencing document properties^^A
+%  \thanks{This module has version
+%    \ltpropertiesversion\ dated \ltpropertiesdate, \copyright\ \LaTeX\
```
[the remark this answers is not in the document]
Comment (u-fischer) [reply base/ltproperties.dtx:42]: öh. where?

File: base/ltproperties.dtx (hunk 6 above)
Comment (u-fischer) [reply base/ltproperties.dtx:777]: no `\G@refundefinedtrue` is command that defines a warning (and yes the name is confusing).

File: base/ltproperties.dtx (hunk 8 above)
Comment (u-fischer) [reply base/ltproperties.dtx:885]: good question ;-).

File: base/ltproperties.dtx (hunk 6 above)
Comment (josephwright) [reply base/ltproperties.dtx:777]: Add a note to the source to highlight this?

File: base/ltproperties.dtx (hunk 9 above)
[the remark this answers is not in the document]
Comment (josephwright) [reply base/ltproperties.dtx:42]: 'The LaTeX Project'

File: base/changes.txt [hunk 10]
```diff
@@ -6,7 +6,13 @@ completeness or accuracy and it contains some references to files that
 are not part of the distribution.
 ================================================================================
 
-2023-08-19 Joseph Wright  <Joseph.Wright@latex-project.org>
+2023-09-05 Ulrike Fischer <Ulrike.Fischer@latex-project.org>
+	* ltproperties.dtx: 
+      First implementation of properties module
+    * ltmiscen.dtx:
+      add test to \enddocument for changes in property values.
```
Comment (muzimuzhi) [inline base/changes.txt:13]: Inconsistent indent characters: 1 tab vs multiple spaces.

It seems most changes entries are indented by 1 tab, excerpt for the one(s) added by myself, which use(s) spaces.

Comment (josephwright) [reply base/changes.txt:13]: Should be tabs: this is (I am told) a standard format from Emacs ...

**Review by josephwright — APPROVED** [review]

**Pull request merged**